### PR TITLE
Start of switchable DKIM Implementation

### PR DESCRIPTION
**Pull request by d00p**
- Rspamd dkim_selector.map dkim_path.map
- Stores dkim_selector in domain table
- Works without enabled DNS Server
- based on feature/rspamd by cs8898

# Description

Please include a summary of the change and which issue is fixed if any. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

* Distribution:
* Webserver:
* PHP:
* etc.etc.:

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: lib/Froxlor/Cron/System/TasksCron.php [hunk 1]
```diff
@@ -57,6 +57,12 @@ public static function run()
 				 * TYPE=2 MEANS TO CREATE A NEW HOME AND CHOWN
 				 */
 				self::createNewHome($row);
+
+			} elseif ($row['type'] == TaskId::REBUILD_DKIM) {
+				/**
+				 * TYPE=41 MEANS THAT SOMETHING IN THE DKIM CONFIG HAS CHANGED.
```
Comment (d00p) [inline lib/Froxlor/Cron/System/TasksCron.php:63]: think you mean "TYPE=14"

File: install/froxlor.sql [hunk 2]
```diff
@@ -390,6 +390,11 @@ INSERT INTO `panel_settings` (`settinggroup`, `varname`, `value`) VALUES
 	('dkim', 'dkim_dkimkeys', 'dkim-keys.conf'),
 	('dkim', 'dkimrestart_command', '/etc/init.d/dkim-filter restart'),
 	('dkim', 'privkeysuffix', '.priv'),
+	('dkim', 'dkim_service_implementation', 'DkimFilter');
```
[the remark this answers is not in the document]
Comment (llucps) [reply install/froxlor.sql]: Sorry if this is not the place, but I understand the dkim-filter is not going away, we're just adding rspamd support. Correct? Will there be config templates in order to setup rspamd?

[the remark this answers is not in the document]
Comment (d00p) [reply install/froxlor.sql]: You need to as that @eis-os - the creater :) I've just created the PR from his repository (we were talking on discord and he had to go)

[the remark this answers is not in the document]
Comment (llucps) [reply install/froxlor.sql]: 👍

[the remark this answers is not in the document]
Comment (eis-os) [reply install/froxlor.sql]: As soon as all parts stabilized, there will some basic config.
Rspamd needs a lot of setup work, specially for security on shared hosting.

The dkim code is extensible, so OpenDKIM could be added as well. 

For rspamd dkim signing config:
https://www.rspamd.com/doc/modules/dkim_signing.html#using-maps

[the remark this answers is not in the document]
Comment (llucps) [reply install/froxlor.sql]: Maybe I'm not getting it, but it seems unless rspamd is setup then this implementation you're doing won't apply. Correct? So, it's not a matter of just changing from dkim-filter to your new rspamd method. Thanks.

[the remark this answers is not in the document]
Comment (d00p) [reply install/froxlor.sql]: @llucps that would always be the case if you switch major services (e.g. webserver from apache to nginx). You will need to reconfigure using the configuration-templates

**Comment by drexlma** [comment]
Hab es mal bei mir getestet. Hilft mir schonmal weiter. danke

Für andere die das geg. auch benötigen: 

**Download des Patches:**
wget https://patch-diff.githubusercontent.com/raw/Froxlor/Froxlor/pull/1007.patch

**Fehlende Files anlegen damit der Patch durchläuft:**
mkdir -p lib/Froxlor/Cron/Dkim/
touch lib/Froxlor/Cron/Dkim/DkimCron.php
touch lib/Froxlor/Cron/Dkim/DkimFilter.php
touch lib/Froxlor/Cron/Dkim/Rspamd.php
mkdir -p lib/Froxlor/UnitTest
touch lib/Froxlor/UnitTest/FroxlorTestCase.php
mkdir tests/Dkim/
touch tests/Dkim/DkimCronTest.php
touch tests/Froxlor/FileDirTest.php
mkdir  lib/Froxlor/Dkim/
touch lib/Froxlor/Dkim/Dkim.php

**Patch testen:**
patch --dry-run -p1 <1007.patch
**Patch anwenden:**
patch  -p1 <1007.patch
   
 **Fehlendes Feld anlegen in der DB**
ALTER TABLE `panel_domains` ADD `dkim_selector` varchar(128) AFTER `dkim_id`;

**Comment by d00p** [comment]
@eis-os is this still being worked on?

**Comment by eis-os** [comment]
I did have a v2 branch, but with the new major version it got quite useless.

https://github.com/Froxlor/Froxlor/commit/5420a99828d7b3c5f690be93e254cfc343daf02a

So, yes, simple close this PR, it won't go any forward.

**Comment by Neustradamus** [comment]
It has been added?

**Pull request closed without merging**